src/sql/mir.rs

```diff
@@ -785,6 +808,142 @@ impl SqlToMirConverter {
+        match *ce {
+            LogicalOp(ref ct) => {
+                let (left, right);
+                match ct.operator {
+                    Operator::And => {
+                        left = self.make_predicate_nodes(
+                            name,
+                            parent.clone(),
+                            &*ct.left,
+                            nc
+                        );
+
+                        right = self.make_predicate_nodes(
+                            name,
+                            left.last().unwrap().clone(),
+                            &*ct.right,
+                            nc + left.len()
+                        );
+
+                        pred_nodes.extend(left.clone());
+                        pred_nodes.extend(right.clone());
+
+                    },
+                    Operator::Or => {
+                        left = self.make_predicate_nodes(
+                            name,
+                            parent.clone(),
+                            &*ct.left,
+                            nc
+                        );
+
+                        right = self.make_predicate_nodes(
+                            name,
+                            parent.clone(),
+                            &*ct.right,
+                            nc + left.len()
+                        );
+
+                        debug!(self.log, "Creating union node for `or` predicate");
+
+                        let last_left = left.last().unwrap().clone();
+                        let last_right = right.last().unwrap().clone();
+                        let union = self.make_union_node(
+                            &format!("{}_u", name),
+                            vec![last_left, last_right]
+                        );
+
+                        pred_nodes.extend(left.clone());
+                        pred_nodes.extend(right.clone());
+                        pred_nodes.push(union);
+                    },
+                    _ => unreachable!("LogicalOp operator is {:?}", ct.operator)
+                }
+            },
+            ComparisonOp(ref ct) => {
+                // currently, we only support filter-like comparison operations, no nested-selections
+                let f = self.make_filter_node(
+                    &format!("{}_f{}", name, nc),
+                    parent,
+                    ct,
+                );
+
+                pred_nodes.push(f);
+            },
+            NegationOp(_) => unreachable!("negation should have been removed earlier"),
+            Base(_) => unreachable!("dangling base predicate"),
+        }
+
+        pred_nodes
+    }
+
+    /// Returns all collumns used in a set of predicates
+    fn cols_in_predicates(
+        &self,
+        ces: Vec<ConditionExpression>
+    ) -> HashSet<Column> {
+        use nom_sql::ConditionExpression::*;
+
+        let mut cols = HashSet::new();
+        for ce in ces {
+            match ce {
```

Comment:
Yes. I changed `qgn.predicates` to be a `Vec<ConditionExpression>` rather than a `Vec<ConditionTree>` because it makes it easier to recursively create nodes in `make_predicate_nodes` (there's less `match` statements and so on). But I guess it has the downside of having to repeat some code to get the columns in each predicate.